tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -576,6 +581,362 @@ INSTANTIATE_TEST_SUITE_P(
+  int M = 1024, N = 512, K = 256;
+  int Ko = num_devices, Ki = K / Ko;
+  std::vector<int> a_shape = {Ko, Ki, M};
+  std::vector<int> b_shape = {Ko, Ki, N};
+
+  TensorView* a = makeContigTensor(3, DataType::Half); // (Ko,Ki,M)
+  TensorView* b = makeContigTensor(3, DataType::Half); // (Ko,Ki,N)
+  // Transpose into TN layout, keep Ko (device axis) as the outermost.
+  TensorView* a_t = transpose(a, 1, 2); // (Ko,M,Ki)
+  TensorView* b_t = transpose(b, 1, 2); // (Ko,N,Ki)
+  TensorView* a_b = broadcast(a_t, {false, false, true, false}); // (Ko,M,b,Ki)
+  TensorView* b_b = broadcast(b_t, {false, true, false, false}); // (Ko,b,N,Ki)
+  TensorView* ab = mul(a_b, b_b); // (Ko,M,N,Ki)
+  TensorView* c0 = sum(ab, {-1}); // (Ko,M,N,r)
+  TensorView* c = sum(c0, {0}); // (r,M,N)
+
+  fusion->addInput(a);
+  fusion->addInput(b);
+  fusion->addOutput(c);
+
+  // Parallelize K on all inputs and intermediates.
+  auto all_sharded_tvs = {a, b, a_t, b_t, a_b, b_b, ab, c0};
+  for (auto tv : all_sharded_tvs) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+    tv->setDeviceMesh(mesh);
+  }
+  c->setDeviceMesh(mesh);
+
+  auto [a_, b_, c_] = getAtenInputOutputs(MmaLayout::NT, M, N, K);
+  a_ = a_.view({Ko, Ki, M});
+  b_ = b_.view({Ko, Ki, N});
+  std::vector<c10::IValue> inputs = {
+      shardTensor(a_, a, communicator->deviceId()),
+      shardTensor(b_, b, communicator->deviceId())};
+
+  MultiDeviceExecutor runtime(
+      std::move(fusion), *communicator, executor_params);
+  auto outputs = runtime.runWithInput(inputs);
+
+  testValidate(
+      runtime.completeFusion(),
+      outputs,
+      inputs,
+      {c_},
+      __LINE__,
+      __FILE__,
+      "",
+      LaunchParams(),
+      getTolerances());
+}
+
+TEST_F(DistributedMatmulTest, LayoutNT_ReduceScatter) {
+  // MmaLayout::NT matmul A(N), B(T), C(T)
+  // A, B are sharded on K. C is sharded on M
+  // Tests local matmul + reduce scatter
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  DeviceMesh mesh = createDeviceMesh();
+
+  // Note: Manually split K and M
+  int M = 1024, N = 512, K = 256;
+  int Ko = num_devices, Ki = K / Ko;
+  int Mo = num_devices, Mi = M / Mo;
+  std::vector<int> a_shape = {Ko, Ki, M};
+  std::vector<int> b_shape = {Ko, Ki, N};
+
+  TensorView* a = makeContigTensor(3, DataType::Half); // (Ko,Ki,M)
+  TensorView* b = makeContigTensor(3, DataType::Half); // (Ko,Ki,N)
+  TensorView* a_t = transpose(a, 1, 2); // (Ko, M, Ki)
+  TensorView* b_t = transpose(b, 1, 2); // (Ko, N, Ki)
+  TensorView* a_b = broadcast(a_t, {false, false, true, false}); // (Ko,M,b,Ki)
+  TensorView* b_b = broadcast(b_t, {false, true, false, false}); // (Ko,b,N,Ki)
+  TensorView* ab = mul(a_b, b_b); // (Ko,M,N,Ki)
+  TensorView* c0 = sum(ab, {-1}); // (Ko,M,N,r)
+  c0 = segment_set(c0);
+  // TODO: Reshape works using sharded sizes. Should use unsharded size to
+  // stay consistent.
```

Comment:
This sounds like a bug. How hard is it to fix that? Indeed it creates inconsistency, e.g., I'd expect `c1->axis(0)->parallelize(DIDx)` somewhere.

Reply:
> it may actually be `c1->axis(1)->parallelize(DIDx)`

This should be `c1->axis(0)->parallelize(DIDx)`, I should have inlucded it in the list of sharded_tvs (will also rename to k_sharded_tvs to not confuse with c which is sharded on M). 
That might fix the reshape problem.

Reply:
> This should be c1->axis(0)->parallelize(DIDx)

You are right -- the reshape is intra-device (so dim 0 should be parallelized on DIDx), and the sum is cross-device (so dim 1 DIDx).